### PR TITLE
Support JRuby

### PR DESCRIPTION
**Pull request by headius**
Tweaks and additions to support JRuby.

### CODE REVIEW

**Comment by headius** [comment]
@hsbt This is mergeable now, but there are some open questions:

* Is the way I handled JRuby's custom Windows logic acceptable?
* Any objections to the addition of the new version.rb file?
* How should I deal with #2?

Once this is released as a gem, JRuby will be able to eliminate its local copy of open3 and use the gem instead.

File: .github/workflows/test-jruby.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Ruby
+      uses: ruby/setup-ruby@v1
+      with:
+        ruby-version: ${{ matrix.ruby }}
+    - name: Install dependencies
+      run: bundle install
+    - name: Run test
+      run: rake test
```
Comment (olleolleolle) [inline .github/workflows/test-jruby.yml:22]: To use the fancy caching, we can:

```suggestion
        ruby-version: ${{ matrix.ruby }}
        bundler-cache: true # 'bundle install' and cache
    - name: Run test
      run: bundle exec rake test
```

Comment (headius) [reply .github/workflows/test-jruby.yml:22]: I have largely been copying these configs from the existing ones. Perhaps a round of PRs to update the other gems, so this doesn't keep getting propagated without the caching?

**Comment by headius** [comment]
This is now complete. I did an additional pass to update and improve the JDK logic from JRuby 9.4, which now errors out for unsupported redirects and reuses the wrapper functions from open3.rb. The JDK version passes all but three tests:

* test_numeric_file_descriptor2 and test_numeric_file_descriptor2 fail due to redirecting streams to a pipe IO, unsupportable using JDK's ProcessBuilder.
* test_pid fails expecting a real PID which we is not exposed by ProcessBuilder.

JRuby's CI runs could add Windows, but these tests would need to be masked out.

Let me know what more I need to do to get this released!

**Comment by headius** [comment]
@hsbt Ping! I believe this is ready to merge, modulo any recent changes to master. See also the change to `jit_support.rb` I detailed in #2 which is in this PR and necessary to pass tests (JRuby cannot support CRuby's JIT APIs).

**Pull request merged**